docs/designs/allreduce.md

```diff
@@ -1,14 +1,215 @@
+in the code below. Each worker continues to perform tasks until there is no new batch available.
+
+```python
+while True:
+    dataset = self._task_data_service.get_dataset()
+    if not dataset:
+        break
+    dataset = dataset.batch(self._minibatch_size).prefetch(1)
+    for dataset_batch in dataset:
+        task = self._task_data_service.get_current_task()
+        process_minibatch_and_report(dataset_batch, task)
+```
+
+If any of the gradients averaging operation fails while the workers are still active and healthy, we simply report the
+failure and continue training on the next batch.
+
+If any of the workers fails, e.g. the pod is accidentally killed, the following steps will be performed:
+
+1. The task that the failed worker was handling will be added back to the task queue.
+1. Since the `AllreduceCommunicator` is aware of the failure. It will reconstruct the communicator and re-assign ranks among
+the existing active workers. The existing workers will then continue to run on the tasks at hand.
+1. The master pod will try to create a new worker pod after a specified restart delay period.
+1. Once the worker pod becomes active and `AllreduceCommunicator` is aware of it, we perform the following steps:
+    1. Lock all existing worker pods.
+    1. Select one of the worker pods to broadcast the current model parameters to all worker pods.
+    1. Release the lock that each worker pod holds.
+    1. All worker pods continue to process training tasks at hand and perform allreduce operations.
+
+Note that since the existing active workers have the exact same copy of the model after the previous allreduce operation completes,
+we can guarantee that the new worker will have the same copy of the model as the ones on other workers once the next
+allreduce operation completes.
+
+#### Training with Evaluation
+
+If the worker encounters any evaluation tasks in the above process, it will evaluate the model directly on the worker once the
+under-going allreduce-based gradients averaging has completed and the model has been updated. The behavior is the same as
+what's described in [model evaluation design doc](model_evaluation.md) except that we are evaluating the model on workers
+instead of on parameter servers. Once an evaluation completes, we send the evaluation result to master for TensorBoard
+service to consume.
+
+#### ElasticDL Embedding Layer
+
+ElasticDL embedding layer is not supported under allreduce-based training since each worker has the exact same copy of
+the model that must fit within each worker's specified resources. Any layers in the model defined via
+``elasticdl.layers.embedding`` will be replaced by `tf.keras.layers.Embedding`.
+
+#### Relevant CLI Arguments
+
+The following CLI arguments are relevant and their behaviors might be changed under allreduce-based training:
+
+* ``--restart_policy``: The pod restart policy when pod crashed.
```

Comment:
Do we have a enum for restart_policy?

Reply:
Not yet but this should be one of the supported values consistent with k8s's pod restart policy: https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#restart-policy

Reply:
I added this in https://github.com/sql-machine-learning/elasticdl/pull/1451.